File: C++/315.count-of-smaller-numbers-after-self.cpp
/*
 * @lc app=leetcode id=315 lang=cpp
 *
 * [315] Count of Smaller Numbers After Self
 *
 * https://leetcode.com/problems/count-of-smaller-numbers-after-self/description/
 *
 * algorithms
 * Hard (38.86%)
 * Likes:    1276
 * Dislikes: 59
 * Total Accepted:    83.8K
 * Total Submissions: 215.6K
 * Testcase Example:  '[5,2,6,1]'
 *
 * You are given an integer array nums and you have to return a new counts
 * array. The counts array has the property where counts[i] is the number of
 * smaller elements to the right of nums[i].
 * 
 * Example:
 * 
 * 
 * Input: [5,2,6,1]
 * Output: [2,1,1,0] 
 * Explanation:
 * To the right of 5 there are 2 smaller elements (2 and 1).
 * To the right of 2 there is only 1 smaller element (1).
 * To the right of 6 there is 1 smaller element (1).
 * To the right of 1 there is 0 smaller element.
 * 
 */
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
private:
    // [start, stop], [start, mid], [mid+1, stop]
    void mergeCountSmaller(vector<pair<int, int>> &nums_index, int start, int stop, vector<int> &res)
    {
        if (stop == start)
            return;
        int mid = (stop - start) / 2 + start;
        int right = mid+1, left = start;
        int temp_cur = 0;
        mergeCountSmaller(nums_index, start, mid, res);
        mergeCountSmaller(nums_index, mid+1, stop, res);
        vector<pair<int,int>> temp(stop-start+1);
        while(left <= mid || right <= stop)
        {
            // if nums_index[left].first == nums_index[right].first, it shouldn't take into count
            if(left> mid || (right <= stop && nums_index[left].first > nums_index[right].first))
            {
                temp[temp_cur] = nums_index[right];
                right++;
                temp_cur++;
            }
            else
            {
                temp[temp_cur] = nums_index[left];
                res[nums_index[left].second] += right - mid - 1;
                left++;
                temp_cur++;
            }
        }
        // inplace_merge(nums_index.begin() + start, nums_index.begin() + mid, nums_index.begin() + end);
        for(int i = start; i <= stop; i++)
        {
            nums_index[i] = temp[i-start];
        }
    }

public:
    vector<int> countSmaller(vector<int> &nums)
    {
        int len = nums.size();
        if (len == 0)
            return vector<int>{};
        vector<int> res(len, 0);
        vector<pair<int, int>> nums_index;
        for (int i = 0; i < len; i++)
            nums_index.push_back(make_pair(nums[i], i));
        mergeCountSmaller(nums_index, 0, len-1, res);
        return res;
    }
};

File: C++/315.count-of-smaller-numbers-after-self.java
import java.util.Arrays;

/*
 * @lc app=leetcode id=315 lang=java
 *
 * [315] Count of Smaller Numbers After Self
 *
 * https://leetcode.com/problems/count-of-smaller-numbers-after-self/description/
 *
 * algorithms
 * Hard (38.86%)
 * Likes:    1276
 * Dislikes: 59
 * Total Accepted:    83.8K
 * Total Submissions: 215.6K
 * Testcase Example:  '[5,2,6,1]'
 *
 * You are given an integer array nums and you have to return a new counts
 * array. The counts array has the property where counts[i] is the number of
 * smaller elements to the right of nums[i].
 * 
 * Example:
 * 
 * 
 * Input: [5,2,6,1]
 * Output: [2,1,1,0] 
 * Explanation:
 * To the right of 5 there are 2 smaller elements (2 and 1).
 * To the right of 2 there is only 1 smaller element (1).
 * To the right of 6 there is 1 smaller element (1).
 * To the right of 1 there is 0 smaller element.
 * 
 */

// import java.util.List;
// import java.util.ArrayList;W

class Solution {
    public List<Integer> countSmaller(int[] nums) {
        // insert sorting
        int len = nums.length;
        if(len == 0)
        {
            return new ArrayList<Integer>();
        }
        int[] temp_res = new int[len];
        List<Integer> temp = new ArrayList<Integer>();
        temp.add(nums[len-1]);

        // now to implement the insert sort
        // from the right to the left
        // the last element is already sorted
        for(int i = len-2; i >=0; i--)
        {
            int left = 0, right = temp.size();
            // f
            while(left < right)
            {
                int mid = left +(right-left)/2;
                if(temp.get(mid)>nums[i])
                    right = mid;
                else left = mid+1;
            }
            temp_res[i] = right;
            temp.add(right, nums[i]);
        }
        
        List<Integer> res = new ArrayList<Integer>(Arrays.asList(temp_res));
        return res;
    }
}

// vector<int> countSmaller(vector<int>& nums)
// {   
//     int len = nums.size();
//     if(len == 0)
//         return vector<int> {};
//     vector<int> res(len);
//     vector<int> temp{nums[len-1]};
//     // begin from the last line
//     for(int i = len-2; i>=0; i--)
//     {
//         int left = 0, right = temp.size();
//         while(left < right)
//         {
//             int mid = left + (right-left)/2;
//             if(temp[mid] >= nums[i])
//                 right = mid;
//             else left = mid+1;
//         }
//         res[i] = right;
//         temp.insert(temp.begin()+right, nums[i]);
//     }
//     return res;
// }
// };

